### ISSUE TITLE
skip resolving images are referenced with their digest

### ISSUE THREAD

**Issue by dprotaso** [k14s/kbld#18]
ie.

```
image: some/image@sha256:12312312...
```

Get's processed by `kbld` and resolves to itself. This also leaves unnecessary annotation on some objects`

ie.

```
kbld.k14s.io/images: |
  - Metas: null
    URL: some/image@sha256:12312312...
```

**Comment by cppforlife** [comment]
thats fair. we should probably remove that.

**Comment by Zebradil** [comment]
Processing resolved image might be useful to verify that the image is actually available.

**Comment by pivotaljohn** [comment]
> Processing resolved image might be useful to verify that the image is actually available.

The guarantee that `kbld` is driving towards is ensuring that references are reliable — that they are digests (read: immutable) rather than tagged references.

The thinking goes that even if we were to check for the presence of the image, it doesn't prevent someone from removing it from the registry/repo moments later... it's not much of a guarantee. And in practice, if we've got digest-articulated coordinates, that URL is likely still legit.

In short, little gain for the cost of using a HTTP call to the registry (thinking of registry rate limiting).

**Comment by Zebradil** [comment]
So far I haven't faced a situation when a digest was removed, but this is still possible.

In some workflows resolved digests are sitting in manifests for a long time (months maybe) and the probability of disappearing that particular digest exactly in the time frame between subsequent invocations of `kbld` and `kubectl` (or `kapp`) is rather small. I personally feel more comfortable when pipelines are falling earlier than later, so that I know about the issue before it gets to production and provokes alerts or unexpected behavior.

Also, there is a way to tell `kbld` to skip resolution of that particular image if it is not desired.

But, to be honest, I don't feel that this is required. I suppose that this scenario might be solved with `kbld` lock files. I just got the idea that it might be not a bug but a feature and thought that this also should be considered.

**Comment by pivotaljohn** [comment]
Thank you for persisting in your perspective, @Zebradil.

One of our core design goals across the Carvel suite is to ensure the operator has confidence in the moves they make with the tools.

I can absolutely envision the situation where I've been handed a bill of materials and want to verify that these "materiel" are present. 👍🏼 

In fact, while `kbld` includes the annotation despite the fact that no real resolution occurred, it hasn't actually verified the presence of that image. 😮 

What do you think about including a flag that modifies this behavior to actually verify the presence of the image in the registry?  Something like `--verify-digest-refs` ?

**Comment by Zebradil** [comment]
I'm not sure if it makes sense to change the default behavior (when `kbld` doesn't actually resolve digest references). So, yes, I think the flag would be a good solution here to let people opt-in.

Thanks for considering this.
